clients/java/src/main/java/io/zeebe/client/impl/ZeebeClientCloudBuilderImpl.java

```diff
@@ -0,0 +1,168 @@
+import java.time.Duration;
+import java.util.Properties;
+
+public class ZeebeClientCloudBuilderImpl implements ZeebeClientCloudBuilder {
+
+  private static final String BASE_ADDRESS = "zeebe.camunda.io";
+  private static final String BASE_AUTH_URL = "https://login.cloud.camunda.io/oauth/token";
+
+  private final ZeebeClientBuilder innerBuilder = new ZeebeClientBuilderImpl();
+
+  private String clusterId;
+  private String clientId;
+  private String clientSecret;
+
+  @Override
+  public ZeebeClientCloudBuilder withClusterId(final String clusterId) {
+    this.clusterId = clusterId;
+    return this;
+  }
+
+  @Override
+  public ZeebeClientCloudBuilder withClientId(final String clientId) {
+    this.clientId = clientId;
+    return this;
+  }
+
+  @Override
+  public ZeebeClientCloudBuilder withClientSecret(final String clientSecret) {
+    this.clientSecret = clientSecret;
+    return this;
+  }
+
+  @Override
+  public ZeebeClientBuilder withProperties(final Properties properties) {
+    return innerBuilder.withProperties(properties);
+  }
+
+  @Override
+  public ZeebeClientBuilder brokerContactPoint(final String contactPoint) {
+    return innerBuilder.brokerContactPoint(contactPoint);
+  }
+
+  @Override
+  public ZeebeClientBuilder gatewayAddress(final String gatewayAddress) {
+    return innerBuilder.gatewayAddress(gatewayAddress);
+  }
+
+  @Override
+  public ZeebeClientBuilder defaultJobWorkerMaxJobsActive(final int maxJobsActive) {
+    return innerBuilder.defaultJobWorkerMaxJobsActive(maxJobsActive);
+  }
+
+  @Override
+  public ZeebeClientBuilder numJobWorkerExecutionThreads(final int numThreads) {
+    return innerBuilder.numJobWorkerExecutionThreads(numThreads);
+  }
+
+  @Override
+  public ZeebeClientBuilder defaultJobWorkerName(final String workerName) {
+    return innerBuilder.defaultJobWorkerName(workerName);
+  }
+
+  @Override
+  public ZeebeClientBuilder defaultJobTimeout(final Duration timeout) {
+    return innerBuilder.defaultJobTimeout(timeout);
+  }
+
+  @Override
+  public ZeebeClientBuilder defaultJobPollInterval(final Duration pollInterval) {
+    return innerBuilder.defaultJobPollInterval(pollInterval);
+  }
+
+  @Override
+  public ZeebeClientBuilder defaultMessageTimeToLive(final Duration timeToLive) {
+    return innerBuilder.defaultMessageTimeToLive(timeToLive);
+  }
+
+  @Override
+  public ZeebeClientBuilder defaultRequestTimeout(final Duration requestTimeout) {
+    return innerBuilder.defaultRequestTimeout(requestTimeout);
+  }
+
+  @Override
+  public ZeebeClientBuilder usePlaintext() {
+    return innerBuilder.usePlaintext();
+  }
+
+  @Override
+  public ZeebeClientBuilder caCertificatePath(final String certificatePath) {
+    return innerBuilder.caCertificatePath(certificatePath);
+  }
+
+  @Override
+  public ZeebeClientBuilder credentialsProvider(final CredentialsProvider credentialsProvider) {
+    return innerBuilder.credentialsProvider(credentialsProvider);
```

Comment:
I'm not sure this method should be implemented. When it is used, the result is actually overwritten when `build` is called. This is also true for `gatewayAddress`.

We could throw an exception, or log a warning to inform the user that they shouldn't use this method. Even better would be to remove this method from the ZeebeClientCloudBuilder, but I guess that would separate interfaces for both the ZeebeClientBuilder and the ZeebeClientCloudBuilder and as well a common interface for both. I don't know if we can find some nice namings then. Do you have any ideas on how to solve this?

Reply:
Hey, maybe we can split the current builder interface with the core part without the auth part and extend it in the current `ZeebeClientBuilder` and in the `ZeebeClientCloudBuilder`? I think this would be the best solution :)

Reply:
Sounds good to me, but let's consider naming for a bit. 

One option is to keep the name `ZeebeClientBuilder` for the core (common/shared) part, then have an extending interface `ZeebeClientCloudBuilder` (adding the cloud methods) and an extending interface named `x` (adding the `credentialsProvider` and `gatewayAddress` methods). It's not yet clear to me what x should be named.

Another option is to give the above x the name `ZeebeClientBuilder` and name the core (common/shared) interface something like `ZeebeClientCommonBuilder` or similar. I personally prefer the first option over this one. 

Extracting part of the interface is a breaking change (although only users that store the builder in a variable would be affected by this change), and the first option only breaks the interface for 2 methods, while the second breaks the interface for all but 2 methods. However, I still want to mention it, because it might be good plan B.

Please consider some ideas for naming the interfaces.

Reply:
Hey @korthout! I've returned to this task, so I don't have a lot of ideas about naming I want to discuss the following:
1. `ZeebeOnPremiesClientBuilder` - it is like opposite of `ZeebeClientCloudBuilder` 🤷🏻‍♂️
2. `ZeebeGeneralClientBuilder` - it is like you can use it both in the cloud and with your on-primes service.
Maybe you have some thoughts/ideas on how to name it? Because I don't like my options 😅

Reply:
I've tried with `ZeebeOnPremiseClientBuilder`, can you take a look? :)

Reply:
Sorry for the wait @aivinog1, I'm still discussing these topic with others team members. I'll have a look at the other changes

Reply:
Just to let you know. Other changes look great! Thanks for that. 

The interface naming part is not looking so nice yet. Having to cast the builder to be able to call `plaintext` looks counter intuitive. It will require some more work, but I'll get back to you about the how as soon as I know more.

Reply:
@korthout Thanks, I wait for some propositions :)

Reply:
@npepinpe had a good idea on how to deal with this while leaving the `ZeebeClientBuilder` as it was. 

Let's allow the user to overwrite defaults. The defaults for the `ZeebeClientCloudBuilder` are to require the `clusterId`, `clientId` and `clientSecret` to be set. And its default behavior is to use these to connect to the cloud cluster. 

However, it can allow to overwrite the default behavior by setting `gatewayAddress` and/or `credentialsProvider` directly. Both then still require either: that the `gatewayAddress` and the `credentialsProvider` are set directly; or that some subset of the `clusterId`, `clientId` and `clientSecret` are set to make the other case still work correctly.

This way the normal ZeebeClientBuilder does not have to change and the ZeebeClientCloudBuilder actually becomes a bit more flexible. I hope my explanation makes sense. Please let me know if it doesn't 😄

Reply:
Hey @korthout! No problem, it was interesting to trying to implement something like this, it was a good experience 👍. I've tried to implement this algorithm as you described, can you check it, please? :)